### PR TITLE
Unify getitem code between Cube, Coord and CellMeasures.

### PR DESCRIPTION
**Pull request by pp-mo**
This is a useful preliminary to replacing lazy/real array data with DataManager in Coords code.

It reduces the indexing code in those 3 areas to suspiciously similar usages, with the bulk of the common code in a new private utility routine `iris.utils._slice_data_with_keys` :  I think that code can usefully go inside `DataManager.__getitem__` in future.

In future, in addition to rebasing all these 3 areas onto DataManager, I would also move the coord points and bounds accessors into `Coord` as generic methods, and make DimCoord extend the generic calls with its extra checks and tweaks.  The new similarities of handling make this attractive, but it is better left undone because (a) it makes this PR simpler and (b) it will be simpler when the underlying data objects are all DataManager-s.

This PR also contains a strictly temporary hack to remove the special-casing of LazyArray data within AuxCoords, by wrapping them as dask arrays instead (`WrappedLazyArray` class and `_wrap_lazy` function in coords.py).  
This should be replaced by the result of https://github.com/SciTools/iris/pull/2518 when that is ready.

### CODE REVIEW

**Comment by pp-mo** [comment]
@bjlittle @dkillick @lbdreyer Please check this out !

**Comment by dkillick** [comment]
> the bulk of the common code in a new private utility routine `iris.utils._slice_data_with_keys` : I think that code can usefully go inside `DataManager.__getitem__` in future.

Out of interest, why did you choose to add this new private utils routine rather than add a `__getitem__` to the `DataManager` class in the first instance?

**Comment by dkillick** [comment]
> why did you choose to add this new private utils routine ... ?

Oh, is the answer "because the routine already existed"? And, as such, was a useful place to host extended functionality...

**Comment by pp-mo** [comment]
> why did you choose to add this new private utils routine 

I just needed a place for common functionality, but I'm expecting it to go into DataManager.
I guess we'll leave the two existing iris.util routines in place as status quo is simplest, though not exactly ideal.

**Comment by pp-mo** [comment]
Rebased.
@dkillick @lbdreyer Please re-check !

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1496,7 +1481,7 @@ def _repr_other_metadata(self):
     @property
     def points(self):
         """The local points values as a read-only NumPy array."""
-        points = self._points.view()
+        points = self._points
         return points
```
Comment (dkillick) [inline lib/iris/coords.py]: You could combine these two lines now, surely?

File: lib/iris/coords.py [hunk 2]
```diff
@@ -1600,9 +1585,6 @@ def _sanitise_array(self, src, ndmin):
         # Ensure the array has enough dimensions.
         # NB. Returns the *same object* if result.ndim >= ndmin
         result = np.array(result, ndmin=ndmin, copy=False)
-        # We don't need to copy the data, but we do need to have our
-        # own view so we can control the shape, etc.
-        result = result.view()
```
Comment (dkillick) [inline lib/iris/coords.py]: Why do we no longer want to make a view of `result`? Is this because we've removed `LazyArray` or is it more in depth than that?

File: lib/iris/cube.py [hunk 3]
```diff
@@ -2168,23 +2161,14 @@ def new_cell_measure_dims(cm_):
...
 
-        if first_slice is not None:
-            data = cube_data[first_slice]
-        else:
-            data = copy.deepcopy(cube_data)
-
-        for other_slice in slice_gen:
-            data = data[other_slice]
+        # Index with the keys, using orthogonal slicing.
+        dimension_mapping, data = iris.util._slice_data_with_keys(
```
Comment (dkillick) [inline lib/iris/cube.py]: Where is `dimension_mapping` used?

File: lib/iris/util.py [hunk 4]
```diff
@@ -696,6 +696,27 @@ def _build_full_slice_given_keys(keys, ndim):
...
+    # NOTE: by using slicing on only one index at a time, this also generally
+    # avoids copying the data, except when a key contains a list of indices.
+    full_slice = _build_full_slice_given_keys(keys, n_parent_dims)
+
+    dims_mapping, slices_iter = column_slices_generator(full_slice,
+                                                        n_parent_dims)
+    for this_slice in slices_iter:
+        data = data[this_slice]
+        if data.ndim > 0 and min(data.shape) < 1:
+            raise IndexError('Cannot index with zero length slice.')
```
Comment (pp-mo) [inline lib/iris/util.py]: Existing code *allows* slicing a cube to 'nothing', like cube[2:2], but *disallows* this for a Coord (with the message here).  So..
```
>>> cube = iris.cube.Cube(np.arange(5))
>>> print cube[1:1]
unknown / (unknown)                 (-- : 0)
>>> cube.add_aux_coord(iris.coords.DimCoord(np.arange(5), long_name='x'), 0)
>>> print cube[1:1]
Traceback (most recent call last):
 . . .
  File "/opt/scitools/environments/experimental/2017_04_27/lib/python2.7/site-packages/iris/coords.py", line 492, in __getitem__
    raise IndexError('Cannot index with zero length '
IndexError: Cannot index with zero length slice.
>>> 
```

There are tests that check for this error, like [here](https://github.com/SciTools/iris/blob/3eb48980cb58a7784d9003d9fbe6f94003c1ce87/lib/iris/tests/test_cdm.py#L610)

So, I've retained it, but the restriction now applies to Cubes as well -- that didn't break any tests :crossed_fingers: 
This effectively disallows it for cubes as well,

File: lib/iris/coords.py (hunk 1 above)
Comment (pp-mo) [inline lib/iris/coords.py]: ok

File: lib/iris/util.py (hunk 4 above)
Comment (dkillick) [inline lib/iris/util.py]: It's a nice outcome of this unification that the same behaviour is now observed for coords and cubes :+1: Might be nice to add a test to confirm the cube on its own also behaves like this now.

File: lib/iris/aux_factory.py [hunk 5]
```diff
@@ -195,7 +195,7 @@ def _nd_bounds(self, coord, dims, ndim):
         nd_shape = [1] * ndim + [coord.nbounds]
         for dim, size in zip(dims, coord.shape):
             nd_shape[dim] = size
-        bounds.shape = tuple(nd_shape)
```
Comment (pp-mo) [inline lib/iris/aux_factory.py]: I think that modifying the view of an array in-place is horrible !

File: lib/iris/coords.py (hunk 2 above)
Comment (pp-mo) [inline lib/iris/coords.py]: My reasoning is that this **ought to be** redundant, now that we no longer support LazyArray-s.

I would generally expect taking a view of an (ordinary) real array to be a redundant operation.
Likewise, our `__init__` codes, and data assignment methods, take the view that a Coord or Cube "owns" it's data, so that should be reasonable.
However, the one place it didn't work was because of assigning array.shape as an in-place modification in aux_factoriy code, e.g. [here](https://github.com/SciTools/iris/pull/2519/files#r114524615).
I think that is highly unexpected, really horrible + shouldn't be done -- so I fixed it.

File: lib/iris/cube.py (hunk 3 above)
Comment (pp-mo) [inline lib/iris/cube.py]: It is used as a free variable in the two methods defined immediately _above_, `new_coord_dims` and `new_cell_measure_dims`.
Not obvious, is it ?!?
I considered moving them to below this point for more clarity, but the minimal change principle won out...

Comment (dkillick) [inline lib/iris/cube.py]: > Not obvious, is it ?!?

In a word, no. I did notice it was being defined in the previous form of the code, but I still wasn't sure it was being used...

Still, :+1: for following the minimum change principle!

File: lib/iris/util.py [hunk 6]
```diff
@@ -696,6 +696,27 @@ def _build_full_slice_given_keys(keys, ndim):
     return full_slice
 
 
+def _slice_data_with_keys(data, keys, n_parent_dims):
+    # Combine _build_full_slice_given_keys and column_slices_generator to index
```
Comment (pp-mo) [inline lib/iris/util.py]: I think this is rather work-in-progress, but I'll clarify it + see if it improves...

Comment (pp-mo) [inline lib/iris/util.py]: This new routine should probably have its own test.
I could write something now, or we could wait until we decide whether this becomes part of DataManager ?

**Comment by pp-mo** [comment]
I think we should maybe have a test for the new function `iris.util._slice_data_with_keys` (which I just changed a bit).
I could do that now, or we could wait until we decide whether it lives inside DataManager ?

File: lib/iris/util.py [hunk 7]
```diff
@@ -696,22 +696,39 @@ def _build_full_slice_given_keys(keys, ndim):
...
-                                                        n_parent_dims)
+def _slice_data_with_keys(data, keys):
+    """
+
+    Index an array-like object with orthogonal keys.
+
+    This enforces an orthogonal interpretation of indexing, which means that
+    both 'real' (numpy) arrays and other array-likes index in the same way,
+    instead of numpy arrays doing 'fancy indexing'.
+
```
Comment (dkillick) [inline lib/iris/util.py]: You could do with a section 'Args:' in the docstring that describes what each of the incoming arguments does.

File: lib/iris/util.py [hunk 8]
```diff
@@ -696,22 +696,39 @@ def _build_full_slice_given_keys(keys, ndim):
...
+    Returns (dim_map, data_region), where :
+
+    *  dim_map (dict) :
+        A dimension map, as for :func:`column_slices_generator`.
```
Comment (dkillick) [inline lib/iris/util.py]: Doesn't the dimension map come _from_ `column_slices_generator`, rather than being used _for_ it? It might also be worth describing what the dimension map does, or how/why it might be used?

**Comment by dkillick** [comment]
These simplifications are worthwhile I think @pp-mo. But I'm afraid that I think we _do_ need some testing for the new `iris.util._slice_data_with_keys` function. Perhaps try and keep the testing quite terse, given that the new function may be only temporary...

File: lib/iris/util.py (hunk 8 above)
Comment (pp-mo) [inline lib/iris/util.py]: > Doesn't the dimension map come from column_slices_generator, rather than being used for it?
> ... It might also be worth describing what the dimension map does, or how/why it might be used?

By "as for" I mean "as it is for" -- so the use is described there, for the other routine.
As this calls the other routine, and it is public, I would much prefer to simply defer to that description.

File: lib/iris/util.py [hunk 9]
```diff
@@ -696,6 +696,44 @@ def _build_full_slice_given_keys(keys, ndim):
     return full_slice
 
 
+def _slice_data_with_keys(data, keys):
+    """
+
```
Comment (lbdreyer) [inline lib/iris/util.py]: Unecessary line?

**Comment by lbdreyer** [comment]
>But I'm afraid that I think we do need some testing for the new iris.util._slice_data_with_keys function. 

Agreed
>Perhaps try and keep the testing quite terse, given that the new function may be only temporary...

but the tests you write could be modified so that they become data manager tests once it's moved there

File: lib/iris/coords.py [hunk 10]
```diff
@@ -1496,8 +1481,7 @@ def _repr_other_metadata(self):
     @property
     def points(self):
         """The local points values as a read-only NumPy array."""
```
Comment (lbdreyer) [inline lib/iris/coords.py]: Why does this say `read-only NumPy array.` ?
```
>>> coord = AuxCoord([1, 2, 3])
>>> print coord.points.flags.writeable
True
>>> print coord._points.flags.writeable
True
>>> coord.points = [4, 5, 6]
>>> print coord.points.flags.writeable
True

```

Comment (lbdreyer) [inline lib/iris/coords.py]: Okay! I see this is DimCoord not AuxCoord:
```
>>> coord = DimCoord([1, 2, 3])
>>> print coord.points.flags.writeable
False
```

File: lib/iris/coords.py (hunk 2 above)
Comment (djkirkham) [inline lib/iris/coords.py]: Isn't the reason for this to stop users messing with the points array 'by the back door'?
```
>>> d = DimCoord([1,2,3,4])
>>> d.points.shape = (2,2)
>>> print d
DimCoord(array([[1, 2],
       [3, 4]]), standard_name=None, units=Unit('1'))
```

**Comment by pp-mo** [comment]
> @lbdreyer but the tests you write could be modified so that they become data manager tests once it's moved there

:+1:

**Comment by pp-mo** [comment]
Hi @lbdreyer @dkillick 
New code with added test.
Hope this addresses all to-date.
More delay --> yet another rebase, I'm afraid.

File: lib/iris/tests/unit/util/test__slice_data_with_keys.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+    def __getitem__(self, keys):
+        return keys
+
+
+# An Indexer object for generating indexing keys in a nice visible way.
+Index = Indexer()
+
+
+class MixinIndexingTest(object):
+    def check(self, shape, keys, expect_call_keys=None, expect_map=None):
```
Comment (djkirkham) [inline lib/iris/tests/unit/util/test__slice_data_with_keys.py]: This check feels a bit strict. As I understand it, the purpose of `_slice_data_with_keys` is to avoid using indices containing multiple tuples, in order to avoid Numpy 'fancy' indexing. So why not just do the following checks:

1. Each index only consists of either non-tuple values, or a single tuple plus any number of `slice(None,None,None)`
2. The resulting array is as expected

That way it gives freedom to tweak the algorithm in the future if necessary, without changing the tests.

Actually what is really being tested is the behaviour of `column_slices_generator`, so these tests shouldn't really be necessary. But the testing for that function is minimal...

Comment (dkillick) [inline lib/iris/tests/unit/util/test__slice_data_with_keys.py]: > what is really being tested is the behaviour of `column_slices_generator`

Or the combination with `_build_full_slice_given_keys`. See also the [note in the docstring](https://github.com/pp-mo/iris/blob/daf80ac690daabb4cc3fe9fb291a5a5fd46f49e1/lib/iris/tests/unit/util/test__slice_data_with_keys.py#L20-L24) for this test module.

> these tests shouldn't really be necessary

We've introduced a new, untested function. Of course there should be tests for it! Especially when the real driver for these tests was the terrible lack of any other testing for anything within this functionality space. Thus adding these tests is a positive step as it gives us some assurance for the first time that this functionality is doing what we thought it was...

You could perhaps argue that these should be integration tests, or alternatively you could say we're testing the unit that is `iris.util._slice_data_with_keys`.

File: lib/iris/tests/unit/util/test__slice_data_with_keys.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+                msg += '\n]'
+                return msg
+
+            self.assertTrue(equal, errmsg.format(showkeys(calls_got),
+                                                 showkeys(expect_call_keys)))
+        if expect_map is not None:
+            self.assertEqual(dim_map, expect_map)
+
+
+class Test_indexing(MixinIndexingTest, tests.IrisTest):
```
Comment (dkillick) [inline lib/iris/tests/unit/util/test__slice_data_with_keys.py]: @pp-mo these tests look good :+1: ... and very thorough!

My only comment, which applies to both test classes here, is that there are very few failure modes. Admittedly there are only two exceptions within all the functions under test here, but are there any other edge cases that need to be considered?

File: lib/iris/tests/unit/util/test__slice_data_with_keys.py (hunk 11 above)
Comment (djkirkham) [inline lib/iris/tests/unit/util/test__slice_data_with_keys.py]: >  Of course there should be tests for it!
You're right, and I'd forgotten that the function also used `_build_full_slice_given_keys`. But still, the checks on the indices really apply to `column_slices_generator`

**Comment by pp-mo** [comment]
Following discussion, I added some purely results-based tests.
The existing mock-ist approach is liable to collapse if we need to modify any of the implementation, but the new test should be robust to refactoring.

File: lib/iris/tests/unit/util/test__slice_data_with_keys.py (hunk 12 above)
Comment (pp-mo) [inline lib/iris/tests/unit/util/test__slice_data_with_keys.py]: This exercises errors raised by the 3 routines covered, but no more.
I'm certainly not trying to cover all cases here, as there are far too many possible forms of indexing (at least in numpy), and who knows what differences between numpy and dask may be lurking ?
I've just picked some known testcases.

**Comment by dkillick** [comment]
I was minded to say I was happy with things as they were _before_ the latest commit, so I'm definitely minded to say I'm happy now! I'll wait to see if we can persuade Travis to be happy and then I think I might merge...
